### DISCUSSION TITLE
Problem with VIPS_BLEND_MODE_SOURCE

### DISCUSSION THREAD

**Discussion by KamikaceDixit** [libvips/libvips#3226] [Q&A] [answered]
Hi! I'm trying to use composite() function with VIPS_BLEND_MODE_SOURCE. I have created a very simple code to test it, but I'm getting problems with it:

```cpp
#include <vips/vips8>

using namespace vips;

int main(int argc, char **argv)
{
    VIPS_INIT(argv[0]);

    VImage blueImage = TSRMultiImage::createEmptyImage(500, 500, "#00B");
    VImage blackImage = TSRMultiImage::createEmptyImage(500, 500, "#000");

    VImage outputImageSource = blueImage.composite(blackImage, VipsBlendMode::VIPS_BLEND_MODE_SOURCE, VImage::option()->set("x", 250)->set("y", 250));
    outputImageSource.write_to_file("outputImageSource.png", VImage::option()->set("Q", 70));

    return (0);
}
```

I'm getting the following result (where the first image has been removed):
![image](https://user-images.githubusercontent.com/43892915/207407448-0465df86-9f3c-44bb-bab3-be29e2583778.png)

And according to the Cairo Documentation (https://www.cairographics.org/operators/), I think it should be something like this:
![image](https://user-images.githubusercontent.com/43892915/207409617-4399e93e-531d-4a92-8c1d-413b29d39afb.png)

I'm getting the last result with VIPS_BLEND_MODE_OVER, but it is only the same as VIPS_BLEND_MODE_SOURCE if alpha = 1. 

I don't know if I'm doing something wrong. 

Thank you a lot!

**Comment by jcupitt** [comment, marked as answer]
Hi @KamikaceDixit,

libvips composite is unbounded, meaning that the stack of images being layered are all the same size, and all pixels are treated in the same way. If an image in the stack is smaller, or has been displaced with `x` and `y`, it is padded and cropped up to full size with 0s (transparent).

For `source`, this means that every pixel in the output is just the second image, and the first image is removed.

**Reply by KamikaceDixit** [reply]
Hi @jcupitt , thank you a lot for your answer,

I understand the first part, but maybe I'm missunderstanding something in the second one.

According Cairo documentation (https://www.cairographics.org/operators/) 

<html>
<body>
<h2 id="source" style="font-family: sans-serif; clear: both; color: rgb(243, 153, 20); margin-left: 46.7812px; margin-right: 46.7812px; font-size: 20px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">CAIRO_OPERATOR_SOURCE</h2><div class="tutright" style="float: right; background-color: rgb(238, 238, 238); border: 1px solid rgb(136, 136, 136); padding: 0.5em; margin-left: 1em; color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><img src="https://www.cairographics.org/operators/source.png" width="160" height="120" class="img" style="border: none; margin: 0px;"></div><p style="margin-left: 93.5781px; margin-right: 93.5781px; color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The second object is drawn as if nothing else were below. Only outside of the blue rectangle the red one is left intact.</p><p style="margin-left: 93.5781px; margin-right: 93.5781px; color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The effect of the<span> </span><code>SOURCE</code><span> </span>operator depends on the interpretation of the source. In cairo, this operator is<span> </span><strong>bounded</strong>.</p>

Resulting alpha (aR) | Resulting color (xR)
-- | --
aA | xA

<br class="Apple-interchange-newline">
</body>
</html>

I understand the first object should be visible according with the image and this: "Only outside of the blue rectangle the red one is left intact."
So I think the pixels of the first image should be replaced for the pixels of the second one in the overlapping area, but the other pixels should remain intact. 

Maybe I am misinterpreting everything and I am using the wrong operation, I need to put the source on top of the destination, regardless the transparecy, replacing the pixels of the destination.

I hope you can help me. Thank you a lot!

**Reply by jcupitt** [reply]
Yes, that's the unbounded part in libvips:  there's no outside the overlap.

For a bounded overlay you'd need to find the common rect (a set intersection on the rectangles I guess), cut that out of both images, composite that area, then insert the result on top of the two source images.

**Reply by jcupitt** [reply]
Or to rephrase it, you could think of a bounded (ie. "has edges") composite as being something you might build on top of an unbounded composite primitive.
